https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>

93c18c4aadd9ce744729bc293ac4af43371c03af 



In [6]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [7]:
# Load the dataset (load remotely or locally)
df = pd.read_csv('iris.csv')

# Output the first 15 rows of the data
print(df.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(df.info())


    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

### Part 1: Analysis

The dataset is a collection of physical measurements of Iris flowers. It contains **150 entries**, each representing a unique flower sample, with specific measurements of four different characteristics related to a flower's sepal and dimensions of the petals. 

**Features**
- `sepal_length`: Length of the sepal in centimeters.
- `sepal_width`: Width of the sepal in centimeters.
- `petal_length`: Length of the petal in centimeters.
- `petal_width`: Width of the petal in centimeters.

These are the quantitative meaasurements of the flower's characteristics that are used to determine its species

**Label**
- `species`: The species of the Iris flower

The column categorizes each flower into one of three classes:

1. Setosa
2. Versicolor
3. Virginica

Each of the names represents a distinct sub-species of Iris flower; the classification task at hand will be to accurately assign new samples to one of these three classes based on the given measurement features.




# Part 2: Split the dataset into train and test

In [8]:
# Take the dataset and split it into our features (X) and label (y)
X = df.drop(columns=['species'])
y = df['species']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, random_state = 400)

# Output the shape of the training and test sets
print("Training set shape:", X_train.shape, y_train.shape)
print("Test set shape:", X_test.shape, y_test.shape)

Training set shape: (135, 4) (135,)
Test set shape: (15, 4) (15,)


# Part 3: Logistic Regression

In [10]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression(max_iter=200)
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = pd.DataFrame({
    'sepal_length': [X_train['sepal_length'].mean()],
    'sepal_width': [X_train['sepal_width'].mean()],
    'petal_length': [X_train['petal_length'].mean()],
    'petal_width': [X_train['petal_width'].mean()]
})

predicted_probabilities = model.predict_proba(sample_datapoint)

print("Predicted Probabilities For Each Class:", predicted_probabilities)

# iii. Report on the score for Logistic regression model, what does the score measure?
score = model.score(X_test, y_test)
cv_scores = cross_val_score(model, X, y, cv=5)

print("Model Accuracy Score:", score)
print("Average Cross-Validation Accuracy:", cv_scores.mean())

# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = model.coef_
intercepts = model.intercept_

print("\nCoefficients:", coefficients)
print("Intercepts:", intercepts)

Predicted Probabilities For Each Class: [[0.05856227 0.92838096 0.01305677]]
Model Accuracy Score: 1.0
Average Cross-Validation Accuracy: 0.9733333333333334

Coefficients: [[-0.40537096  0.96671479 -2.42461787 -1.04631712]
 [ 0.52832694 -0.42185106 -0.19618086 -0.87957121]
 [-0.12295598 -0.54486373  2.62079873  1.92588834]]
Intercepts: [  9.29245484   2.33792946 -11.6303843 ]


### Part 3: Logistic Regression Analysis

The accuracy score measures the proportion of correctly classified samples in the test set. The Logistic Regression model achieved an accuracy score of **1.0** indicating that the model correctly classified all test samples. A perfect accuacy score suggests that the model was very effective in learning the patterns in the data to distinguish between classes.

After going through the entire assignment, all of my models achieved an accuracy score of **1.0**, making me skeptical that I may have gotten a lucky seed for my train/test split. I tried a variety of seeds, all of which seemingly had no effect on the model accuracy score. So, to further compare the models, I implemented a cross-validation with 5 folds. 

The average cross-validation score represents the mean accuracy across multiple splits, or folds, of the dataset, which represents a more accurate estimate of the model's general performance. The Logistic Regression model achieved an average cross-validation accuracy of **97.33%**, meaning that on average, the model correctly classified approximately 97.33% of samples across the 5 different subsets of the data. This is a high score, suggesing that the model consistently generalizes well to unseen data.


### Decision Boundaries

#### Boundary between Class 1 and Class 2
$$
\begin{align*}
0 &= -0.9337 \cdot \text{sepal\_length} \\
&+ 1.3886 \cdot \text{sepal\_width} \\
&- 2.2284 \cdot \text{petal\_length} \\
&- 0.1667 \cdot \text{petal\_width} \\
&+ 6.9545
\end{align*}
$$

#### Boundary between Class 1 and Class 3
$$
\begin{align*}
0 &= -0.2824 \cdot \text{sepal\_length} \\
&+ 1.5116 \cdot \text{sepal\_width} \\
&- 5.0454 \cdot \text{petal\_length} \\
&- 2.9722 \cdot \text{petal\_width} \\
&+ 20.9228
\end{align*}
$$

#### Boundary between Class 2 and Class 3
$$
\begin{align*}
0 &= 0.6513 \cdot \text{sepal\_length} \\
&+ 0.1230 \cdot \text{sepal\_width} \\
&- 2.8170 \cdot \text{petal\_length} \\
&- 2.8055 \cdot \text{petal\_width} \\
&+ 13.9683
\end{align*}
$$


# Part 4: Support Vector Machine

In [61]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svm_model = SVC(probability=True)
svm_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
predicted_probabilities_svm = svm_model.predict_proba(sample_datapoint)
print("Predicted Probabilities For Each Class:", predicted_probabilities_svm)

# iii. Report on the score for the SVM, what does the score measure?
svm_score = svm_model.score(X_test, y_test)
svm_cv_scores = cross_val_score(svm_model, X, y, cv=5)

print("Model Accuracy Score:", svm_score)
print("Average Cross-Validation Accuracy:", svm_cv_scores.mean())

Predicted Probabilities For Each Class: [[0.02243889 0.97226602 0.00529508]]
Model Accuracy Score: 1.0
Average Cross-Validation Accuracy: 0.9666666666666666


### Part 4: SVM Analysis

The SVC model achieved an accuracy score of **1.0**, indicating that it correctly classified all samples in the test set. A perfect score suggests the model effectively separates the classes in this split.

The model also achieved an average cross-validation score of **96.67%**, again derived from 5 different folds. This is a high score, further confirming that the SVC model consistently generalizes well.


# Part 5: Neural Network

In [ ]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier(max_iter = 800, random_state = 400)
mlp_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
predicted_probabilities_mlp = mlp_model.predict_proba(sample_datapoint)
print("Predicted Probabilities For Each class:", predicted_probabilities_mlp)

# iii. Report on the score for the Neural Network, what does the score measure?
mlp_score = mlp_model.score(X_test, y_test)
mlp_cv_scores = cross_val_score(model, X, y, cv=5)

print("Neural Network model score (accuracy):", mlp_score)
print("Average Cross-Validation Accuracy:", mlp_cv_scores.mean())

# iv: Experiment with different options for the neural network, report on your best configuration
# Define different configurations for experimentation
hidden_layer_sizes_options = [(10,), (50,), (100,), (50, 50)]
activation_options = ['relu', 'tanh', 'logistic']
alpha_options = [0.0001, 0.001, 0.01]

best_score = 0
best_params = {}

# Experiment with different configurations
for hidden_layer_sizes in hidden_layer_sizes_options:
    for activation in activation_options:
        for alpha in alpha_options:
            
            # Initialize the MLP model with current parameters
            mlp_model = MLPClassifier(
                hidden_layer_sizes=hidden_layer_sizes,
                activation=activation,
                alpha=alpha,
                max_iter=2000,
                random_state=400
            )

            mlp_model.fit(X_train, y_train)
            mlp_cv_scores = cross_val_score(mlp_model, X, y, cv=5)
            mlp_avg_cv_score = mlp_cv_scores.mean()
            
            # Update the best score and parameters if the current configuration scores better
            if mlp_avg_cv_score > best_score:
                best_score = mlp_avg_cv_score
                best_params = {
                    'hidden_layer_sizes': hidden_layer_sizes,
                    'activation': activation,
                    'alpha': alpha
                }

print("\nBest Neural Network Configuration:")
print("Hidden Layer Sizes:", best_params['hidden_layer_sizes'])
print("Activation Function:", best_params['activation'])
print("Alpha Regularization:", best_params['alpha'])
print("Best Accuracy Score (Avg. Cross-Validation):", best_score)

Predicted Probabilities For Each class: [[9.58124597e-03 9.90337060e-01 8.16938184e-05]]
Neural Network model score (accuracy): 1.0
Average Cross-Validation Accuracy: 0.9733333333333334

Best Neural Network Configuration:
Hidden Layer Sizes: (10,)
Activation Function: logistic
Alpha Regularization: 0.0001
Best Accuracy Score (Avg. Cross-Validation): 0.9866666666666667


### Part 5: NN Analysis

The MLP Classifier model achieved a perfect accuracy score of **1.0** on the test set, indicating it correctly classified all samples. This result suggests the model effectively separates the classes in this particular split, with no misclassifications.

Additionally, the model achieved an average cross-validation score of **97.33%** over 5 folds.

**Best Neural Network Configuration**:

For the experimental portion of this analysis, I explored several configurations, including varying hidden layer sizes, activation functions, and regularization values. I tested single hidden layers with different neuron counts, as well as a two-layer setup. I also compared relu, tanh, and logistic activation functions, and experimented with different alpha values to control regularization. To ensure reliability, I evaluated each configuration using the average cross-validation score, given that the test set accuracy was consistently 1.0 across configurations.

Th network configuration that achieved the highest accuracy score was a single hidden layer with **10** neurons, a `logistic` activation function, and an `alpha` regularization value of **0.0001**. It achieved an average accuracy score of **98.67%** across the 5 folds, showing a slight improvement over the inital cross-validation score, confirming its optimality for this dataset.

# Part 6: K-Nearest Neighbors

In [56]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
knn_model = KNeighborsClassifier(n_neighbors=5)
knn_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
predicted_probabilities_knn = knn_model.predict_proba(sample_datapoint)
print("Predicted probabilities For Each Class:", predicted_probabilities_knn)

# iii. Report on the score for kNN, what does the score measure?
knn_score = knn_model.score(X_test, y_test)
knn_cv_scores = cross_val_score(knn_model, X, y, cv=5)

print("k-Neighbors Classifier model score (accuracy):", knn_score)
print("Cross-Validation Accuracies for each fold:", knn_cv_scores)
print("Average Cross-Validation Accuracy:", knn_cv_scores.mean())

Predicted probabilities For Each Class: [[0. 1. 0.]]
k-Neighbors Classifier model score (accuracy): 1.0
Cross-Validation Accuracies for each fold: [0.96666667 1.         0.93333333 0.96666667 1.        ]
Average Cross-Validation Accuracy: 0.9733333333333334


### Part 6: k-NN Analysis

The k-NN model achieved an accuracy score of **1.0**, indicating that it correctly classified all samples in the test set. A perfect score suggests the model effectively separates the classes in this split.

The model also achieved an average cross-validation score of **97.33%**, again derived from 5 different folds. This is a high score, further confirming that the k-NN model generalizes well.

# Part 7: Conclusions and takeaways

Overall, each model performed exceptionally well on the dataset, all achieving perfect accuracy scores, with minimal variation across different folds. I experimented over a few different seeds (42, 123, 456, 400) and the accuracy scores were all consistently 1.0. As a result, the comparison for "best performance" was based on the Average Cross-Validation Accuracy.

Ultimately, the MLP Classifier with a single hidden layer of **10** neurons, `logistic` activation, and **0.0001** alpha achieved the highest average cross-validation score of **98.67%**. This model likely performed best because the configuration struck a balance between model complexity and generalization ability.

The uniformly high performance across all models suggests that the dataset is well-structured and likely linearly separable, allowing simpler models such as logistic regression to achieve results comparable to the more complex neural networks. This exercise allowed me to visualize first hand the concept behind Occam's razor, showing me the importance of choosing a model based on the dataset's characteristics rather than defaulting to complexity.